src/Prometheus/Storage/APCng.php

```diff
@@ -0,0 +1,813 @@
+                }
+            }
+            if ($loopCatcher <= 0) {
+                throw new RuntimeException('Caught possible infinite loop in ' . __METHOD__ . '()');
+            }
+        }
+    }
+
+    /**
+     * @param mixed[] $data
+     * @throws RuntimeException
+     */
+    public function updateCounter(array $data): void
+    {
+        // Taken from https://github.com/prometheus/client_golang/blob/66058aac3a83021948e5fb12f1f408ff556b9037/prometheus/value.go#L91
+        $valueKey = $this->valueKey($data);
+        $done = false;
+        $loopCatcher = self::CAS_LOOP_RETRIES;
+        while (!$done && $loopCatcher-- > 0) {
+            $old = apcu_fetch($valueKey);
+            if ($old !== false) {
+                $done = apcu_cas($valueKey, $old, $this->toBinaryRepresentationAsInteger($this->fromBinaryRepresentationAsInteger($old) + $data['value']));
+            } else {
+                apcu_add($valueKey, 0);
+                apcu_store($this->metaKey($data), json_encode($this->metaData($data)));
+                $this->storeLabelKeys($data);
+            }
+        }
+        if ($loopCatcher <= 0) {
+            throw new RuntimeException('Caught possible infinite loop in ' . __METHOD__ . '()');
+        }
+    }
+
+    /**
+     * @param array<string> $metaData
+     * @param string $labels
+     * @return string
+     */
+    private function assembleLabelKey(array $metaData, string $labels): string
+    {
+        return implode(':', [ $this->prometheusPrefix, $metaData['type'], $metaData['name'], $labels, 'label' ]);
+    }
+
+    /**
+     * Store ':label' keys for each metric's labelName in APCu.
+     *
+     * @param array<mixed> $data
+     * @return void
+     */
+    private function storeLabelKeys(array $data): void
+    {
+        // Store labelValues in each labelName key
+        foreach ($data['labelNames'] as $seq => $label) {
+            $this->addItemToKey(implode(':', [
+                $this->prometheusPrefix,
+                $data['type'],
+                $data['name'],
+                $label,
+                'label'
+            ]), isset($data['labelValues']) ? $data['labelValues'][$seq] : ''); // may not need the isset check
+        }
+    }
+
+    /**
+     * Ensures an array serialized into APCu contains exactly one copy of a given string
+     *
+     * @return void
+     * @throws RuntimeException
+     */
+    private function addItemToKey(string $key, string $item): void
+    {
+        // Modify serialized array stored in $key
+        $arr = apcu_fetch($key);
+        if (false === $arr) {
+            $arr = [];
+        }
+        if (!array_key_exists($item, $arr)) {
+            $arr[$this->encodeLabelKey($item)] = 1;
+            apcu_store($key, $arr);
+        }
+    }
+
+    /**
+     * @deprecated use replacement method wipeStorage from Adapter interface
+     *
+     * @return void
+     */
+    public function flushAPC(): void
```

Comment:
We could remove this, as this is a new storage engine

Reply:
Good point, I've removed it.